### ISSUE TITLE
duplicated ALERTs in HA setup on VM - AWS EC2 ASG (auto scaling groups) - AWS Timestream

### ISSUE THREAD

**Issue by zumby** [grafana/grafana#64805]
**What happened**:

Grafana Alerts are duplicated despite the Grafana is configured to use HA Alerting.

**What you expected to happen**:
No duplicates should happen for HA Alerting Grafana setup.

**How to reproduce it (as minimally and precisely as possible)**:

1. We use AWS EC2 to run Grafana servers. EC2 under ASG (auto-scaling groups).
* on each boot, the EC2 instance gets the info about 'peers' and loads the local IP addresses into GF_UNIFIED_ALERTING_HA_* variables that are passed into `/etc/sysconfig/grafana-server`
```
GF_UNIFIED_ALERTING_ENABLED=true
GF_UNIFIED_ALERTING_HA_PUSH_PULL_INTERVAL=200ms
GF_UNIFIED_ALERTING_HA_PEERS='10.10.10.2:9094,10.10.10.1:9094'
GF_UNIFIED_ALERTING_HA_LISTEN_ADDRESS='10.10.10.1:9094'
GF_UNIFIED_ALERTING_HA_ADVERTISE_ADDRESS='10.10.10.1:9094'
```
2. We use **AWS Timestream** as the **datasource** for dashboards and alerts. Here is the example of the query:
```sql
SELECT
    ID,
    device_name as DEVICE,
    CREATE_TIME_SERIES(
        time,
        (
            CASE
                WHEN upper(measure_value :: varchar) = 'UNEXPECTED' THEN 1
                ELSE 0
            END
        )
    ) as Value
FROM
    $ __database.$ __table
WHERE
    $ __timeFilter
    AND measure_name = '$__measure'
GROUP BY
    device_name,
    ID
```
This query awaits for a status 'UNEXPECTED' to appear, and if it appears, it reports 1, otherwise 0. (As there is no string alerts supported by Grafana)
*Spoiler*: the issue happens for all the queries and all the alerts we have. It's NOT caused by **this** specific query for sure.
* we use Reduce and Math expressions in alert setup. Alert condition is Math ($B >= 1) .

3. Our alerts are being *duplicated* despite it seems that the HA configuration is OK.
* alert status history - has duplicates
<img width="702" alt="image" src="https://user-images.githubusercontent.com/26290616/225297676-d434433c-5a6a-496a-ab03-8eb7374ff13d.png">

* alert notifications - has duplicates
<img width="482" alt="image" src="https://user-images.githubusercontent.com/26290616/225293456-6aa187dd-e9b2-465e-8e91-0a1e0d2dfea7.png">

**Anything else we need to know?**:
* the issue does not happen when there is only ONE EC2 instance of Grafana server, thus we think it's the HA setup issue or some bug.
* we've tried changing the HA config in multiple ways and made many tests with no luck
* during the tests in the past we've observed the **missingseries** status in the history, but with the latest config - not anymore

**Environment**:
- Grafana version: OSS v9.4.3
- Data source type & version: AWS Timestream
- OS Grafana is installed on: amazon linux 2

We are not sure if it's a configuration issue or some bug - thus we would appreciate any help or advice on the matter. 
Thank you.

**Comment by zumby** [comment]
found related post: https://learn.microsoft.com/en-us/answers/questions/1139541/duplicate-alerts-firing-in-azure-managed-grafana

**Comment by usmangt** [comment]
Hi @zumby,

Thanks for opening this issue.

As you found a related issue, so did you find the solution from there (as https://github.com/grafana/grafana/issues/53990 is closed) OR this is an ongoing issue for you?

Kindly let us know so that we can set the status of this issue in the correct order.

**Comment by zumby** [comment]
hey @usmangt 

Thanks for response,

this is an **ongoing** issue. The HA Alerting setup seems to have the solution for Kubernetes but not for bare VMs (multiple VMs HA setup)

The closest thread i could found is this: https://learn.microsoft.com/en-us/answers/questions/1139541/duplicate-alerts-firing-in-azure-managed-grafana and it was not resolved as I understand

**Comment by zumby** [comment]
Not sure if this helps, but found some strange `Error:<nil>` in debug logs
```log
logger=ngalert.scheduler
rule_uid=NHfglea
org_id=1
version=4
attempt=0
now=2023-03-19T18:15:00Z
t=2023-03-19T18:15:04.755747647Z
level=debug msg="Alert rule evaluated"
results="[{Instance:DEVICE=DEV-SIMULATION1, UUID=12345 State:Normal Error:<nil> Results:map[] Values:map[B:{Var:B Labels:DEVICE=DEV-SIMULATION1, UUID=12345 Value:0xc0011f9680} C:{Var:C Labels:DEVICE=DEV-SIMULATION1, UUID=12345 Value:0xc0011f9690}] EvaluatedAt:2023-03-19 18:15:00 +0000 UTC EvaluationDuration:4.755725762s EvaluationString:[ var='B' labels={DEVICE=DEV-SIMULATION1, UUID=12345} value=0 ], [ var='C' labels={DEVICE=DEV-SIMULATION1, UUID=12345} value=0 ]}]"
duration=121.120807ms
```

**Comment by grobinson-grafana** [comment]
Hi! When running Grafana in HA the alerts are duplicated N times, where N is the number of replicas. This is intended behaviour as the design of Grafana HA is to share nothing.

However to prevent receiving multiple notifications for the same alerts Grafana in HA mode will deduplicate notifications. You should receive just 1 notification despite having N duplicate alerts, one per replica.

I hope this helps!

**Comment by zumby** [comment]
@grobinson-grafana but the problem is exactly that we do receive more than 1 notification. the 'deduplication' does not happen.

**Comment by grobinson-grafana** [comment]
Can you check that each Grafana pod can communicate with the other pods on port 9094?

**Comment by grobinson-grafana** [comment]
Before we look at the issue further, do you know if all Grafana pods can talk UDP and TCP with the other pods on port 9094?

**Comment by zumby** [comment]
I regret to tell, but the issue is back again in v10.1.1
We are forced to make Grafana NON-HA again and wait for the fix.
@grobinson-grafana 
- 9094 is accessible between Grafana VMs
- we had no such issue on 9.5.1 and 9.5.2

**Comment by armandgrillet** [comment]
Can you check your Grafana metrics to confirm the Grafana VMs are communicating to each other? `grafana_alerting_nflog_gossip_messages_propagated_total` and `alertmanager_cluster_members` and `alertmanager_cluster_failed_peers` and `alertmanager_cluster_health_score` are the interesting ones.

**Comment by zumby** [comment]
It's going to be a bit tough. But let's try. @armandgrillet 

EC2 instances are VMs in AutoScaling group.
They are both behind the same AWS Load Balancer pointing to grafana.domain.com


GF_UNIFIED_ALERTING_HA_PEERS are set automatically during instance boot.
How: it makes a lookup for 'other' instances in the same AutoScaling group and add their IPs to GF_UNIFIED_ALERTING_HA_PEERS.

There is no analog to Kubernetes 'service' that we can use, so thus we supply the IPs of instances directly. Unless we could somehow utilise the Load Balancer address and open 9094 there, though not sure if it is going to work as the networking becomes a bit complicated if we do so.

So now we have the setup is like this:
EC2 instance #1 (10.10.10.236):
```
GF_UNIFIED_ALERTING_HA_PEERS=
10.10.10.236:9094,
10.10.16.129:9094,
10.10.17.169:9094,
10.10.17.247:9094
```

```
EC2 instance #2 (10.10.17.247):
GF_UNIFIED_ALERTING_HA_PEERS=
10.10.16.113:9094,
10.10.17.42:9094,
10.10.17.247:9094
```

And here is what's been requested:

**EC2 instance #1 /metrics** (unknown which one exactly as load balancer distributes the requests 50/50)
```
grafana_alerting_nflog_gossip_messages_propagated_total 2
alertmanager_cluster_members 2
alertmanager_cluster_failed_peers 4
alertmanager_cluster_health_score 0
```



**EC2 instance #2 /metrics** (unknown which one exactly as load balancer distributes the requests 50/50)
```
grafana_alerting_nflog_gossip_messages_propagated_total 0
alertmanager_cluster_members 2
alertmanager_cluster_failed_peers 2
alertmanager_cluster_health_score 0
```

**Comment by grobinson-grafana** [comment]
Thanks! These metrics suggest that the Grafana instances cannot talk to each other. Have you checked whether this is the case? You should be able to observe whether TCP/UDP traffic is being gossiped between the instances.

**Comment by zumby** [comment]
@grobinson-grafana the ports are not the problem. we are sure about this 100% and we were able to verify the communication between EC2 instances on 9094 TCP/UDP

The problem is, as we think is the way config is set for each EC2 in auto-scaling group.
Each new instance gets the value of GF_UNIFIED_ALERTING_HA_PEERS variable upon the start. But the 'old' or 'existing' instances know nothing about new instances. So their config remains outdated. 
This way we end up with few Grafana servers in auto-scaling group that have **different** configs. 

I suppose, in Kubernetes this is somehow resolved through the configmap that has the 'service' internal DNS name. So it is kind of a dynamic config. Unlike with AWS Auto-scaling - we cannot put any similar address there, because, if we for example put the Load Balancer address there, the network routing will become strange - instances will start talking to each other through the 'internet' instead of internal network. In general, we have not tried this as we consider this is not a good idea.
ex. config for K8s:
```
[unified_alerting]
enabled = true
ha_listen_address = "${POD_IP}:9094"
ha_peers = "grafana-alerting.grafana:9094"
ha_advertise_address = "${POD_IP}:9094"
ha_peer_timeout = 15s
```
(https://grafana.com/docs/grafana/latest/alerting/set-up/configure-high-availability/)

So, at this point we are unsure how should we proceed. Looks like we are forced to move either to Kubernetes for self-managed Grafana or Grafana Cloud or AWS Managed Grafana.
For now we just keep our Grafana NON-HA to avoid duplication of alerts.

**Comment by zumby** [comment]
@grobinson-grafana  @armandgrillet 

In addition to above, we have tried forcibly changing the Grafana config and setting the proper HA PEERS manually.

Here are the results:

EC2 instance 1 (10.10.10.236):
config:
`GF_UNIFIED_ALERTING_HA_PEERS='10.10.10.236:9094, 10.10.10.247:9094'
`
metrics:
```
grafana_alerting_nflog_gossip_messages_propagated_total 11
alertmanager_cluster_members 2
alertmanager_cluster_failed_peers 0
alertmanager_cluster_health_score 0
```

EC2 instance 2 (10.10.10.247):
config:
`GF_UNIFIED_ALERTING_HA_PEERS='10.10.10.236:9094, 10.10.10.247:9094'
`
metrics:
```
grafana_alerting_nflog_gossip_messages_propagated_total 16
alertmanager_cluster_members 2
alertmanager_cluster_failed_peers 0
alertmanager_cluster_health_score 0
```

You can see that the configs at both servers are **identical**. Also you can see that metrics reporting everything is OK, except the strange difference in `grafana_alerting_nflog_gossip_messages_propagated_total` count

This is the fresh result of an alert duplication from today with the above config.
<img width="495" alt="image" src="https://github.com/grafana/grafana/assets/26290616/1b23737b-bbb5-4a13-9055-240e84baee5a">

So, bottom line - this has something to do with one of:
- recent changes
- alert labels
- unclear configuration issue

**Comment by gregarndt** [comment]
We've been watching this Issue as we're currently having this problem on 9.4.1, so if they are related, it's not specific to v10.

**Comment by grobinson-grafana** [comment]
Hi! :wave:

> Each new instance gets the value of GF_UNIFIED_ALERTING_HA_PEERS variable upon the start. But the 'old' or 'existing' instances know nothing about new instances. So their config remains outdated.

As long as the new Grafana instances join with at least one existing Grafana via GF_UNIFIED_ALERTING_HA_PEERS, then all instances will learn about each other. It shouldn't be a problem that instances have different configurations. The only case where I think this could be a problem is if one of the oldest instances restarts and none of the instances in GF_UNIFIED_ALERTING_HA_PEERS are reachable anymore.

I think the best way to understand what's happening is to share your configuration (i.e. group_wait, group_interval, repeat_interval, peer_timeout), and to compare the debug logs from each of your Grafana instances.

The most recent screenshot doesn't look like duplication, but looks like the alert is flapping (which is something different). However, I'd need to see debug logs to be certain one way or another.

**Comment by zumby** [comment]
@grobinson-grafana im not sure what exactly should i look for in the debug logs.
As for the group_wait, group_interval, repeat_interval - this is the setup we have, it's pretty basic. Honestly, I did not bother too much setting it up so there might be something 'wrong' there
<img width="753" alt="image" src="https://github.com/grafana/grafana/assets/26290616/e369032a-15a6-4233-94dc-2ecacfb765e7">

**Comment by fapeliberty** [comment]
Hello @zumby 

Don't know if it can help you but I have exactly the same configuration as you. AWS ASG with HA pair configuration. At start, i didn't enable the unified alerting system. And everything worked correctly, i got one notification by alert.

I wanted to use the : "Continue matching subsequent sibling nodes", to create a slack channel that can be a bucket for all notifications without using the default root tree option.

![image](https://github.com/grafana/grafana/assets/85160876/60d57fe4-f8f3-4c60-995b-90c1cbc3ea23)

But it started to trigger two notifications in each channel, je crois qui a une statso i found your issue and the related documentation for unified alert system.

So i follow the same configuration :
```
[unified_alerting]
enabled = true
ha_peers = monitoring-production-nodes.awstack:9094
ha_advertise_address = {{ node-ip }}:9094
```

monitoring-production-nodes.awstack is just a DNS record how store nodes IPs as memberlist, I used lifecycle hook to trigger a lambda to add or remove a node from the list.

After restarting, i got this log, tell me the setup worked :
`level=info msg="gossip settled; proceeding"`

After testing, i don't received multiple notifications.

Grafana version : 10.1.1
Only TCP 9094 open for gossip

**Comment by armandgrillet** [comment]
We do not have sharding in OSS Grafana, the deduplication is happening at the Alertmanager level. Can you double check your routing configuration to check that your alerts are not being sent to the same contact point twice? The status history will indeed show duplicates with your configuration but you should be able to not receive duplicate notifications.

**Comment by armandgrillet** [comment]
Closing as no reply for 2 weeks, feel free to reopen once you have more information.
